lib/puppet/functions/eval.rb

```diff
@@ -0,0 +1,47 @@
+# Evaluates a string containing Puppet Language source.
+# The primary intended use case is to combine eval with
+# Deferred to enable evaluating arbitrary code on the agent side
+# when applying a catalog.
+#
+# @example Using `eval`
+#
+# ```puppet
+# eval("\$x + \$y", { 'x' => 10, 'y' => 20}) # produces 30
+# ```
+#
+# Note the escaped `$` characters since interpolation is unwanted.
+#
+# ```puppet
+# Deferred('eval' ["\$x + \$y", { 'x' => 10, 'y' => 20})] # produces 30 on the agent
+# ```
+#
+# This function can be used when there is the need to format or transform deferred
+# values since doing that with only deferred values can be difficult to construct
+# or impossible to achieve when a lambda is needed.
+#
+# @example Evaluating logic on agent requiring use of "filter"
+#
+# ```puppet
+# Deferred('eval', "local_lookup('key').filter |\$x| { \$x =~ Integer }")
+# ```
+#
+# To assert the return type - this is simply done by calling `assert_type`
+# as part of the string to evaluate.
+#
+# @example
+# ```puppet
+# eval("assert_type(Integer, \$x + \$y))", { 'x' => 10, 'y' => 20})
+# ```
+# @since 6.1.0
+#
+Puppet::Functions.create_function(:eval, Puppet::Functions::InternalFunction) do
+  dispatch :eval_puppet do
+    compiler_param
+    param 'String', :code
+    optional_param 'Hash[String, Any]', :variables
+  end
+
+  def eval_puppet(compiler, code, variables = {})
+    compiler.in_local_scope(variables) { compiler.evaluate_string(code) }
```

Comment:
What are the security implications of evaluating arbitrary strings?

Reply:
Over and beyond what you can already do with EPP or custom Ruby functions called via Deferred (which can do anything), or the types and providers that you have in your environment, or custom facts? If so: Nothing whatsoever that you could not already do - in fact, somewhat more safe since in puppet language you do not have access to everything in Ruby. 

Would it be possible to attack this in a new way? No, it is text in files in manifests as opposed to text in .rb files in an env / modulepath - so same attack vector as for all the existing artefacts.

Could you write something that gets  input from some source and evaluate arbitrary code? Not without explicitly doing something that stupid, and you can do that today with any of the other existing features that pushes logic to be evaluated on the agent.

Reply:
Sure, custom functions running on the agent could do anything, depending on how they are written. But they don't evaluate arbitrary input. The proposed `eval` function takes a string. Where does that string come from? Can it contain data from an unprivileged user allowing the user to escalate privileges when the agent running as root next runs?

Reply:
Hi @hlindberg, just following up on this- can you please answer Josh's question here?

Reply:
There is absolutely no difference between the input to this eval function and the input to any other function - both can be given a string read from any source but that would be the choice of the user. In the case of `eval` it evaluates the string that it is given in the manifest at compile time - the typical use would be that it is a literal string doing some small piece of puppet logic. The `eval` function in itself does not pull in a string from somewhere - they user would have to do that in the manifest, just like with all other functions.

Since the `eval` only evaluates puppet logic, there are no instructions in puppet that makes it possible to escalate privileges that it could execute. There would have to be malicious functions for that, and if those were to exist they are the problem, not eval since you could call them agent side. (Same problem with malicious type/provider implementations).

Reply:
> There is absolutely no difference between the input to this eval function and the input to any other function - both can be given a string read from any source but that would be the choice of the user.

Yes, but it's how the function validates its input that matters. For example, the `generate` function expects an absolute path to a command. The parameter could be a malicious value, so it [validates its input](https://puppet.com/docs/puppet/latest/function.html#generate):

> This function will attempt to protect you from malicious generator calls (e.g., those with ‘..’ in them), but it can never be entirely safe. No subshell is used to execute generators, so all shell metacharacters are passed directly to the generator, and all metacharacters are returned by the function. 

So this is correctly rejected:

```
$ bx puppet apply -e 'notice(generate("/usr/bin/whoami && /bin/cat /etc/passwd"))'
Error: Evaluation Error: Error while evaluating a Function Call, Generators can only contain alphanumerics, file separators, and dashes (line: 1, column: 8) on node
```

> Since the eval only evaluates puppet logic

Ok, I missed this earlier. This is good, so it's not arbitrary ruby code. So this is correctly rejected:

```
$ bx puppet apply -e 'notice(eval("File.read(\"/etc/passwd\")"))'
Error: Evaluation Error: Unknown function: 'read'. (line: 1, column: 10)
```

But this is not:

```puppet
notice(eval("inline_template('<%= File.read(\"/etc/passwd\") %>')"))
```

```
bx puppet apply manifest.pp | head -10
Notice: Scope(Class[main]): ##
# User Database
#
# Note that this file is consulted directly only when the system is running
# in single-user mode.  At other times this information is provided by
# Open Directory.
#
# See the opendirectoryd(8) man page for additional information about
# Open Directory.
##
```

AFAIK there's no way to validate the input to `eval`